### ISSUE TITLE
High level bindings?

### ISSUE THREAD

**Issue by corngood** [expipiplus1/vulkan#7]
Have you put any thought into high level bindings?  Perhaps a hand written one to build on the mechanically generated low level bindings, sort of like OpenGL vs OpenGLRaw?

I started [sketching something out](https://github.com/corngood/vulkan-hl/blob/master/app/Main.hs).  It's going well, but I'm now starting to deal with things that will need a lot of boilerplate, like VkFormat, which I'd rather didn't end up [like this](https://github.com/haskell-opengl/OpenGL/blob/master/src/Graphics/Rendering/OpenGL/GL/PixelFormat.hs).  TemplateHaskell can't deal with the pattern synonyms in Graphics.Vulkan.Core, and going back to the xml spec seems like a bad idea.

**Comment by expipiplus1** [comment]
I think that high level bindings could be done in a couple of different ways:

 - Generate them all by hand, the boilerplate only has to be written once (ignoring future tweaks) and things such as VkFormat could even be produced with `sed` or something similar.

 - Generate them automatically from the spec, using more information than just the C types. There is lots of potentially parsable information such as [range restrictions](https://github.com/KhronosGroup/Vulkan-Docs/blob/1.0/src/spec/vk.xml#L412) or [optional values](https://github.com/KhronosGroup/Vulkan-Docs/blob/1.0/src/spec/vk.xml#L862) which should be included in the type system wherever possible (In fact, there's probably some really cool stuff one could do with indexed monads to make sure everything (or some things) are [mapped when they should be](https://github.com/KhronosGroup/Vulkan-Docs/blob/1.0/src/spec/vk.xml#L582)).

I imagine that the final result will include a combination of these ideas, given the size of the spec it'll take a very long time to get a complete high level implementation written completely by hand. I think a good middle ground would be to create a package along the lines of `vulkan-helpers` which wrap the functions in `vulkan` to make them feel more idiomatic (remove the need for the user to call `alloca` and pass around gross types such as `Ptr`).

I agree it's annoying that template Haskell can't deal with the pattern synonyms, but I think they're still the neatest way of implementing "open data types" such as the Vulkan enumerations.

**Comment by expipiplus1** [comment]
I'm writing a parser for those xml `usage` nodes now, hopefully that'll be useful.

**Comment by corngood** [comment]
I'm going to look at renaming things to make them more haskell-ish (e.g. removing prefixes and suffixes that are encoded in the module name, removing underscores, etc).  What do you think about doing this in Graphics.Vulkan, instead of in a separate high-level binding?

That could be a good first step, as a lot of the types would be useful in the high level binding.  For example, Graphics.Vulkan.Format as an open enum that can be extended.  I think it would be better to use it directly rather than try to wrap it with a closed type.

**Comment by vlsh** [comment]
I'd rather just have a solid tested set of raw bindings first, and extensions are still not done. If you are going to add validation on top of it, you have to remember most calls in Vulkan are meant to be used from multiple threads and none of them are synchronized, which means you'd probably end up having to add a layer of useless locks, this is not just OpenGL anymore.

**Comment by expipiplus1** [comment]
The extensions which don't depend upon 'foreign' types (such as `HWND`) are included in 1.6 thanks to #8. The plan for extensions which do depends on 'foreign' types is to have them in their own packages such as `vulkan-win32` which can depend on both `vulkan` and `win32`.

There's almost certainly a nice spot in the design space where more invariants are encoded in the type system than we are currently doing in `vulkan`.

Even with adding a very high level binding on top of this which has all kinds of (potentially unoptimal) locking  involved I still think that Vulkan will be more pleasant to program against than OpenGL; it's nice to have properly typed enumerations and more predictable performance.

**Comment by vlsh** [comment]
> The plan for extensions which do depends on 'foreign' types is to have them in their own packages such as vulkan-win32 which can depend on both vulkan and win32
I'm sorry, but I hate this idea. Win32 package is a really huge dependency, and Vulkan only needs a few type aliases. I think it would be much simpler to just redefine whatever basic types an extension might need as type aliases and keep them in that same package. Whatever types extension may require they will always be just some primitive C type, and Haskell doesn't even support passing structs by value. Whatever function takes, it's always either a pointer or a number. So to Haskell it's just some type alias and type aliases just have to agree in type, but not in name. Forcing everyone to use Win32 package is a bad idea, it's not an absolute authority on Windows programming and it's not even complete. Windows programming is a very complex and constantly changing activity, APIs come and go, and these packages are not keeping up. And for other applications, like games and graphics applications Win32 is an overkill. I rewrote all my Windows API bindings by hand, just from reading MSDN, and it only took 300 lines, mostly definitions of Storable. When I will be using Vulkan I will just pass same primitive types that it expects, it's not type safe in Haskell sense, but it is type safe in C sense. Anyway I'll probably work on this myself soon enough, I don't have much of a choice.

> Even with adding a very high level binding on top of this which has all kinds of (potentially unoptimal) locking involved I still think that Vulkan will be more pleasant to program against than OpenGL

I was just hoping such experiments could be subject for a another separate package based on this, not everyone might have the same ideas what high-level means, but here we can at least agree based on vk.xml file!

**Comment by expipiplus1** [comment]
> Win32 package is a really huge dependency

I was under the impression that the `Win32` package is almost certain to be included when compiling on Windows, both `directory` and `time` depend on it, each of which have other 1000 packages directly depending on them. 

> I was just hoping such experiments could be subject for a another separate package based on this, not everyone might have the same ideas what high-level means, but here we can at least agree based on vk.xml file!

Sure, I'm not set on putting them in with this package. Hopefully some of the generator code can be useful to them though.

**Comment by headprogrammingczar** [comment]
expipiplus1 is correct. Here's a list of packages that directly depend on `directory` and `time` (and not even a complete list)

http://packdeps.haskellers.com/reverse/directory
http://packdeps.haskellers.com/reverse/time

Even `Cabal` depends on Win32.

More to the point, using the types provided in Win32 is the correct approach. They define exactly what's needed for foreign code to work properly on Windows.

**Comment by vlsh** [comment]
Here's vk.xml defining external dependencies:

```
<type requires="windows.h" name="HINSTANCE"/>
<type requires="windows.h" name="HWND"/>
```
And here's how to define them in Haskell:
```
type HINSTANCE = Ptr ()
type HWND = Ptr ()
```
That's it! You need a library for that?

**Comment by jmcarthur** [comment]
A dependency you already have installed is about as lightweight as you can get. I don't understand the objection to depending on Win32.

**Comment by corngood** [comment]
Win32 is pretty straightforward, but what about  supporting: xlib, xcb, Mir, Wayland, and android?

**Comment by expipiplus1** [comment]
If Hackage doesn't already include definitions for the types Vulkan depends upon then I have no objection to defining these ourselves in the `vulkan-mir` package pending proper bindings appearing.

**Comment by corngood** [comment]
I wrote that on my phone, and it didn't come across very clearly.  I meant that as an argument against including windowing system support in the core package.  While referencing win32 could work just like it does in `directory`, i.e.:

```
    if os(windows)
        build-depends: Win32
```

That's only because os == windowing system in this case.  It gets a whole lot more complicated when other operating systems can support any combination of four windowing systems.

So you either need a bunch of flags for windowing system support (yuck), or separate packages.

**Comment by expipiplus1** [comment]
I totally agree. I think everyone agrees multiple packages is a good idea.

**Comment by expipiplus1** [comment]
This issue seemed to have gotten a little sidetracked.

The new release (sorry it took so long) defines everything locally with no local dependencies:

`Ptr ()` for the Win32 handles (the same type as the `Win32` Package)
https://github.com/expipiplus1/vulkan/blob/master/src/Graphics/Vulkan/Extensions/VK_KHR_win32_surface.hs#L78

`Word32` and opaque data for xcb 
https://github.com/expipiplus1/vulkan/blob/master/src/Graphics/Vulkan/Extensions/VK_KHR_xcb_surface.hs#L83

It's probably not worth splitting hairs over, as it's a very minor friction to anyone using the library, basically wrapping a couple of calls with `coerce` to get them to interoperate with whatever library they're using.

I tried to get the definitions here to match the C definitions in the target libraries or to match other representations on Hackage.

(I've made a new high level discussion here: https://github.com/expipiplus1/vulkan/issues/28, please pitch in!)
